src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Import/ImportFormDataProvider.php

```diff
@@ -26,122 +26,99 @@
-    {
-        $this->session = $session;
+    public function __construct(
+        FileFinder $importFileFinder,
+        SessionInterface $session
+    ) {
         $this->importFileFinder = $importFileFinder;
-        $this->request = $requestStack->getCurrentRequest();
+        $this->session = $session;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function getData()
+    public function getData(ImportConfigInterface $importConfig)
     {
-        // If import entity is available in the query - grab it and preselect in the form,
-        // otherwise - take it from the session.
-        if (null !== $this->request && $this->request->query->has('import_type')) {
-            $entity = $this->request->query->get('import_type');
-        } else {
-            $entity = $this->session->get('entity');
-        }
-
         return [
-            'csv' => $this->getSelectedFile(),
-            'entity' => $entity,
-            'iso_lang' => $this->session->get('iso_lang'),
-            'separator' => $this->session->get('separator', ImportSettings::DEFAULT_SEPARATOR),
-            'multiple_value_separator' => $this->session->get(
-                'multiple_value_separator',
-                ImportSettings::DEFAULT_MULTIVALUE_SEPARATOR
-            ),
-            'truncate' => $this->session->get('truncate', false),
-            'regenerate' => $this->session->get('regenerate', false),
-            'match_ref' => $this->session->get('match_ref', false),
-            'forceIDs' => $this->session->get('forceIDs', false),
-            'sendemail' => $this->session->get('sendemail', true),
+            'csv' => $this->getSelectedFile($importConfig),
+            'entity' => $importConfig->getEntityType(),
+            'iso_lang' => $importConfig->getLanguageIso(),
+            'separator' => $importConfig->getSeparator(),
+            'multiple_value_separator' => $importConfig->getMultipleValueSeparator(),
+            'truncate' => $importConfig->truncate(),
+            'regenerate' => $importConfig->skipThumbnailRegeneration(),
+            'match_ref' => $importConfig->matchReferences(),
+            'forceIDs' => $importConfig->forceIds(),
+            'sendemail' => $importConfig->sendEmail(),
+            'type_value' => [],
         ];
     }
 
     /**
-     * Data is persisted into session,
-     * so when user comes from 2nd import step to 1st one, data is still saved.
-     *
-     * @param array $data
-     *
-     * @return array
+     * {@inheritdoc}
      */
     public function setData(array $data)
     {
         $errors = [];
 
-        if (!isset($data['csv']) || empty($data['csv'])) {
+        if (empty($data['csv'])) {
             $errors[] = [
                 'key' => 'To proceed, please upload a file first.',
                 'domain' => 'Admin.Advparameters.Notification',
                 'parameters' => [],
             ];
+        } else {
+            $this->session->set('csv', $data['csv']);
```

Comment:
@matks that's the only service in which session is injected and used now, but it is required in order to preselect the configuration values if user goes back from import step 2 to step 1. Is that reasonable? :)

Reply:
It is reasonable 😄 

However I'm curious: I see you introduced `ImportFormDataProviderInterface`. That is a good idea because this feature is so complex and specific. But since you created an Interface for this case, why not inject the session in ` getData()` and `setData()` ?

Reply:
Well the interface does not limit a developer to use the session, it's only this particular implementation that requires the session. 
There already are two different import form data providers, one for step 1 (`ImportFormDataProvider`) and one for step 2 (`ImportDataConfigurationFormDataProvider`) and both implement the same `ImportFormDataProviderInterface`, while only one of them depends on the session.